src/ensembl/src/content/app/entity-viewer/state/general/entityViewerGeneralActions.ts

```diff
@@ -52,17 +53,35 @@ export const setDataFromUrl: ActionCreator<ThunkAction<
   Action<string>
 >> = (params: EntityViewerParams) => (dispatch, getState: () => RootState) => {
   const state = getState();
-  let { genomeId } = params;
-  const activeGenomeId = getEntityViewerActiveGenomeId(state);
-  const activeEntityId = getEntityViewerActiveEnsObjectId(state);
-  if (!genomeId) {
+  const { genomeId: genomeIdFromUrl } = params;
+
+  let activeGenomeId = getEntityViewerActiveGenomeId(state);
+  const activeEntityId = getEntityViewerActiveEnsObjectId(state) || undefined;
+
+  const entityIdForUrl = activeEntityId
+    ? buildFocusIdForUrl(activeEntityId)
+    : undefined;
+
+  if (!genomeIdFromUrl && !activeGenomeId) {
     dispatch(setDefaultActiveGenomeId());
-    genomeId = getEntityViewerActiveGenomeId(state) as string;
-  } else if (genomeId !== activeGenomeId) {
-    dispatch(setActiveGenomeId(genomeId));
-    dispatch(fetchGenomeData(genomeId));
+    activeGenomeId = getEntityViewerActiveGenomeId(state) as string;
+  } else if (!genomeIdFromUrl && activeGenomeId) {
+    const newUrl = urlHelper.entityViewer({
+      genomeId: activeGenomeId,
+      entityId: entityIdForUrl
+    });
+    dispatch(replace(newUrl));
+  } else if (genomeIdFromUrl && genomeIdFromUrl !== activeGenomeId) {
+    dispatch(setActiveGenomeId(genomeIdFromUrl));
+    dispatch(fetchGenomeData(genomeIdFromUrl));
     // TODO: when backend is ready, entity info may also need fetching
-  } else {
+  } else if (activeGenomeId && entityIdForUrl) {
+    const newUrl = urlHelper.entityViewer({
+      genomeId: activeGenomeId,
+      entityId: entityIdForUrl
+    });
+    dispatch(replace(newUrl));
```

Comment:
This is very scary, because this is the routing concern that GeneView component is handling; and now there are two places in code that can send conflicting commands to the router. This code branch shouldn't be necessary (and in fact was causing bugs in my PR https://github.com/Ensembl/ensembl-client/pull/324. I deleted it in that PR.

Reply:
I tried removing this but it doesn't seem to retain the state when we switch between species.

Can you please try the following:
- Select BRCA2 for humans
- Switch to a different species
- Go back to Human

The example links are shown again.

Reply:
@msmanoj  I have pushed a commit addressing this issue to your branch.

Briefly, it removes the redirect logic from `entityViewerGeneralActions` and puts it in the `EntityViewer` component. With this change, I expect that the routing hook in GeneView is not going to compete with the routing logic in the action. Before this change, the `replace` function both in the action and in the GeneView could fire; but with this logic moved to EntityViewer component, the GeneView will not even be mounted unless the `replace` function in the EntityViewer fires; and if  GeneView has mounted, the `replace` function in EntityViewer will not fire.